### PR TITLE
add DisableSdsServer flag

### PR DESCRIPTION
**Pull request by KenHuffmanAtNice**
**Please provide a description of this PR:**

The istio-agent/agent.go code checks whether the security.WorkloadIdentitySocketPath exists before starting the Istio SDS server to create that socket.

When using helm to install Istio and SPIRE, the order in which SPIRE and Istio are started is non-deterministic, so this may cause the Istio agent to incorrectly create the socket if SPIRE is started after Istio. It is a kubernetes anti-pattern to force a specific start order during deployment. 

This adds an optional environment variable (like the other boolean agent options) that inhibits the startup of the SDS server by agent.go even when that socket doesn’t yet exist because SPIRE hasn't gotten around to it.

### CODE REVIEW

File: pilot/cmd/pilot-agent/options/options.go [hunk 1]
```diff
@@ -152,4 +152,7 @@ var (
 	exitOnZeroActiveConnectionsEnv = env.Register("EXIT_ON_ZERO_ACTIVE_CONNECTIONS",
 		false,
 		"When set to true, terminates proxy when number of active connections become zero during draining").Get()
+
+	disableSdsServerEnv = env.Register("DISABLE_SDS_SERVER", false,
```
Comment (lei-tang) [inline pilot/cmd/pilot-agent/options/options.go:156]: In the usage comment of this flag, can you describe under which scenarios, this flag should be set as true?

Comment (KenHuffmanAtNice) [reply pilot/cmd/pilot-agent/options/options.go:156]: Fixed in commit: [5ea155585df4f50eaff0ab5bb8b370f1fccd8b8f](https://github.com/istio/istio/pull/41440/commits/5ea155585df4f50eaff0ab5bb8b370f1fccd8b8f)

**Review by lei-tang — APPROVED** [review]

**Review by lei-tang — COMMENTED** [review]
Consider adding a release notes for the new flag.

**Comment by KenHuffman** [comment]
> Consider adding a release notes for the new flag.

I checked other environmental flags in that class and **none** of them have release notes.

**Comment by costinm** [comment]
Release note can be "Fixed SPIRE integration bug" - but it's your call. Would be useful to raise awareness. 

On the flag I have few concerns:
1. We have 2 SDS servers - one in Istiod for Ingress certs, the other one in agent. Maybe a more explicit name (DISALBE_AGENT_SDS) ? Also - we are not disabling SDS - just the built-in SDS server in agent. 
2. I don't know if we still  have any option to completely disable SDS - and use cert files. If you can do a quick search,
and make sure this does not interfere. If we no longer have it - we may want to bring it back at some point.
3. The intent is for agent to auto-detect its environment to reduce user settings, while still allowing user to explicitly
override. Unfortunately over time we accumulated many flags that few people still understand, and there are different 
code paths unlikely to work as expected, since multiple flags determine different paths. I am wondering if for your 
case, the intent is not really to disable SDS - but to specify the explicit choice to use SPIRE ( or some other SDS-over-UDS
based server ) as CA provider. 

How about using CA_ADDR=sds://PATH_TO_UDS_SOCKET ? The CA_ADDR is intended to indicate where
to get the certs - normally it's a host:port but could be extended easily. It is also something we should consider
adding to ProxyConfig, next to discoveryAddress. 

Long term we hope certs will be provided by platform - either as SDS ( path or even address ) or files, and all 
the env variables we use to be gone.

**Comment by KenHuffmanAtNice** [comment]
> Release note can be "Fixed SPIRE integration bug" - but it's your call. Would be useful to raise awareness.

Release notes created

> On the flag I have few concerns:
> 
> 1. We have 2 SDS servers - one in Istiod for Ingress certs, the other one in agent. Maybe a more explicit name (DISALBE_AGENT_SDS) ? Also - we are not disabling SDS - just the built-in SDS server in agent.

Environment variable renamed to DISABLE_AGENT_SDS as you suggested, but I am welcome to a different name since the intent is not to disable the SDS api.

> 2. I don't know if we still  have any option to completely disable SDS - and use cert files. If you can do a quick search,
>    and make sure this does not interfere. If we no longer have it - we may want to bring it back at some point.

I do not wish to disable SDS within Istio, I just want ensure the SPIRE agent will handle SDS requests on that socket.

> 3. The intent is for agent to auto-detect its environment to reduce user settings, while still allowing user to explicitly
>    override. Unfortunately over time we accumulated many flags that few people still understand, and there are different
>    code paths unlikely to work as expected, since multiple flags determine different paths. I am wondering if for your
>    case, the intent is not really to disable SDS - but to specify the explicit choice to use SPIRE ( or some other SDS-over-UDS
>    based server ) as CA provider.

The if statement I’m adding skips the initSdsServer() logic that creates the secret handler. The existing code already skips this function when SPIRE has already created the socket. I’m just guaranteeing that behavior in the case where the SPIRE agent hasn’t started yet, which can happen in a helm deploy.
 
> How about using CA_ADDR=sds://PATH_TO_UDS_SOCKET ? The CA_ADDR is intended to indicate where to get the certs - normally it's a host:port but could be extended easily. It is also something we should consider adding to ProxyConfig, next to discoveryAddress.

I don’t need to make the path to the uds socket configurable. It is a well-known path.

File: pilot/cmd/pilot-agent/options/options.go [hunk 2]
```diff
@@ -152,4 +152,7 @@ var (
 	exitOnZeroActiveConnectionsEnv = env.Register("EXIT_ON_ZERO_ACTIVE_CONNECTIONS",
 		false,
 		"When set to true, terminates proxy when number of active connections become zero during draining").Get()
+
+	disableAgentSdsEnv = env.Register("DISABLE_AGENT_SDS", false,
```
Comment (KenHuffmanAtNice) [inline pilot/cmd/pilot-agent/options/options.go:156]: Should this be put in MeshConfig instead?

Comment (hzxuzhonghu) [reply pilot/cmd/pilot-agent/options/options.go:156]: How about `ENABLE_CUSTOM_SDS` as we are not really disabling sds

**Comment by KenHuffmanAtNice** [comment]
I've come up with a better solution to control the timing between Istio and SPIRE by adding an initContainer within an IstioOperator overlay. This code is no longer necessary for my purposes.

**Pull request closed without merging**